File: jzoj/3939.cpp
#include <bits/stdc++.h>

const int N = 100000 + 10, LIM = 1000000 + 10;

int n, a[N];
std::vector<int> prime;

void preprocessing() {
  static bool flag[LIM];
  prime.clear();
  for (int i = 2; i < LIM; ++i) {
    if (!flag[i]) prime.push_back(i);
    for (int j = 0; j < prime.size() && i * prime[j] < LIM; ++j) {
      flag[i * prime[j]] = true;
      if (i % prime[j] == 0) break;
    }
  }
}

int min[LIM], sub[LIM], minid[LIM], subid[LIM];

std::vector<int> list, cnt;

#define sqr(x) ((x) * (x))

void factor(int x) {
  list.clear();
  cnt.clear();
  for (int i = 0; i < prime.size() && sqr(prime[i]) <= x; ++i) {
    if (x == 1) break;
    if (x % prime[i] == 0) {
      list.push_back(prime[i]);
      cnt.push_back(0);
      while (x % prime[i] == 0) {
        x /= prime[i];
        cnt.back()++;
      }
    }
  }
  if (x > 1) list.push_back(x), cnt.push_back(1);
}

inline void update(int num, int val, int curid) {
  if (val < min[num]) {
    sub[num] = min[num];
    subid[num] = minid[num]; 
    min[num] = val;
    minid[num] = curid;
  } else if (val < sub[num]) {
    sub[num] = val;
    subid[num]  = curid;
  }
}

void dfs(int dep, int num, int rem, int id) {
  if (dep == list.size()) {
    update(num, rem, id);
    return;
  }
  if (cnt[dep]) {
    cnt[dep]--;
    dfs(dep, num * list[dep], rem - 1, id);
    cnt[dep]++;
  }
  dfs(dep + 1, num, rem, id);
}

int best, id;

void solve(int dep, int num, int rem, int curid) {
  if (dep == list.size()) {
    int tmp = rem + (minid[num] == curid ? sub[num] : min[num]);
    int tid = (minid[num] == curid) ? subid[num] : minid[num];
    if (tmp < best || (tmp == best && tid < id)) {
      best = tmp;
      id = tid;
    }
    return;
  }
  if (cnt[dep]) {
    cnt[dep]--;
    solve(dep, num * list[dep], rem - 1, curid);
    cnt[dep]++;
  }
  solve(dep + 1, num, rem, curid);
}

int main() {
/*#ifndef ONLINE_JUDGE
  freopen("in.txt", "r", stdin);
  freopen("out.txt", "w", stdout);
#endif*/
  scanf("%d", &n);
  preprocessing();
  memset(min, 0x3f, sizeof min);
  memset(sub, 0x3f, sizeof sub);
  for (int i = 1; i <= n; ++i) {
    scanf("%d", a + i);
    factor(a[i]);
    dfs(0, 1, std::accumulate(cnt.begin(), cnt.end(), 0), i);
  }
  for (int i = 1; i <= n; ++i) {
    factor(a[i]);
    id = n + 1, best = INT_MAX;
    solve(0, 1, std::accumulate(cnt.begin(), cnt.end(), 0), i);
    printf("%d\n", id);
  }
  return 0;
}

File: OI/jzoj/3939.pas
const	maxn=100000+10;lim=1000000+10;limm=1000;
var	n,best,id,i,j,tmp:longint;
	a:array[1..maxn] of longint;
	prime,list,cnt:array[0..limm] of longint;
	min,sub,minid,subid:array[0..lim] of longint;
        flag:array[1..lim] of boolean;
procedure push(x:longint);
begin
	inc(prime[0]);
	prime[prime[0]]:=x;
end;
procedure prepar;
var     i:longint;
begin
	fillchar(flag,sizeof(flag),false);
	for i:=2 to lim-1 do begin
		if not flag[i] then push(i);
		j:=1;
		while (j<=prime[0]) and (i*prime[j]<lim) do begin
			flag[i*prime[j]]:=true;
			if i mod prime[j]=0 then break;	//i的倍数以过滤
			inc(j);
		end;
	end;
end;
procedure factor(x:longint);
var	i:longint;
begin
	fillchar(list,sizeof(list),0);
	fillchar(cnt,sizeof(cnt),0);
	i:=1;
	while (i<=prime[0]) and (sqr(prime[i])<=x) do begin
		if x=1 then break;
		if x mod prime[i]=0 then begin
			inc(list[0]);
			list[list[0]]:=prime[i];
			inc(cnt[0]);
			cnt[cnt[0]]:=0;
			while x mod prime[i]=0 do begin
				x:=x div prime[i];
				inc(cnt[cnt[0]]);
			end;
		end;
		inc(i);
	end;
	if (x>1) then begin
		inc(list[0]);
		list[list[0]]:=x;
                inc(cnt[0]);
		cnt[cnt[0]]:=1;
	end;
end;
procedure update(num,val,curid:longint);
begin
	if (val<min[num]) then begin
		sub[num]:=min[num];
		subid[num]:=minid[num];
		min[num]:=val;
		minid[num]:=curid;
	end else if (val<sub[num]) then begin
		sub[num]:=val;
		subid[num]:=curid;
	end;
end;
procedure dfs(dep,num,rem,id:longint);
begin
	if (dep>list[0]) then begin
		update(num,rem,id);
		exit;
	end;
	if (cnt[dep]<>0) then begin
		dec(cnt[dep]);
		dfs(dep,num*list[dep],rem-1,id);
		inc(cnt[dep]);
	end;
	dfs(dep+1,num,rem,id);
end;

procedure solve(dep,num,rem,curid:longint);
var	tmp,tid,t:longint;
begin
	if dep>list[0] then begin
		if minid[num]=curid then begin
			t:=sub[num];
			tid:=subid[num];
	       	end else begin
			t:=min[num];
			tid:=minid[num];
		end;
		tmp:=rem+t;
		if (tmp<best) or ((tmp=best) and (tid<id)) then begin
			best:=tmp;
			id:=tid;
		end;
                exit;
	end;
	if cnt[dep]<>0 then begin
		dec(cnt[dep]);
		solve(dep,num*list[dep],rem-1,curid);
		inc(cnt[dep]);
	end;
	solve(dep+1,num,rem,curid);
end;
begin
        //assign(input,'data.in');reset(input);
	readln(n);
	prepar;
        fillchar(min,sizeof(min),$3f);
        fillchar(sub,sizeof(sub),$3f);
	for i:=1 to n do begin
                readln(a[i]);
		factor(a[i]);
		tmp:=0;
		for j:=1 to cnt[0] do inc(tmp,cnt[j]);
		dfs(1,1,tmp,i);
	end;
	for i:=1 to n do begin
		factor(a[i]);
		id:=n+1;
		best:=maxlongint;
		tmp:=0;
		for j:=1 to cnt[0] do inc(tmp,cnt[j]);
		solve(1,1,tmp,i);
		writeln(id);
	end;
end.
